draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -3774,32 +3772,90 @@ Input: u, an element of F.
 Output: (x, y), a point on E.
 
 Constants:
-1.  c1 = -B / A
-2.  c2 = -1 / Z
+1.  c1 = Z / h
 
 Steps:
-1.  tv1 = Z * u^2
-2.  tv2 = tv1^2
-3.   x1 = tv1 + tv2
-4.   x1 = inv0(x1)
-5.   e1 = x1 == 0
-6.   x1 = x1 + 1
-7.   x1 = CMOV(x1, c2, e1)    # If (tv1 + tv2) == 0, set x1 = -1 / Z
-8.   x1 = x1 * c1 # x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
-9.  gx1 = x1^2
-10. gx1 = gx1 + A
-11. gx1 = gx1 * x1
-12. gx1 = gx1 + B             # gx1 = g(x1) = x1^3 + A * x1 + B
-13.  x2 = tv1 * x1            # x2 = Z * u^2 * x1
-14. tv2 = tv1 * tv2
-15. gx2 = gx1 * tv2           # gx2 = (Z * u^2)^3 * gx1
-16.  e2 = is_square(gx1)      # If is_square(gx1)
-17.   x = CMOV(x2, x1, e2)    #   then  x = x1,  else  x = x2
-18.  y2 = CMOV(gx2, gx1, e2)  #   then y2 = gx1, else y2 = gx2
-19.   y = sqrt(y2)
-20.  e3 = sgn0(u) == sgn0(y)  # Fix sign of y
-21.   y = CMOV(-y, y, e3)
-22. return (x, y)
+1.  tv1 = u^2
+2.  tv1 = Z * tv1
+3.  tv2 = tv1^2
+4.  tv2 = tv2 + tv1
+5.  tv3 = tv2 + 1
+6.  tv3 = B * tv3
+7.  tv4 = CMOV(Z, -tv2, tv2 != 0)
+8.  tv4 = A * tv4
+9.  tv2 = tv3^2
+10. tv5 = tv4^2
+11. tv5 = A * tv5
+12. tv2 = tv2 + tv5
+13. tv2 = tv2 * tv3
+14. tv6 = tv4^3
+15. tv5 = B * tv6
+16. tv2 = tv2 + tv5
+17.   x = tv1 * tv3
+18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
+19.   y = c1 * tv1
+20.   y = y * u
+21.   y = y * y1
+22.   x = CMOV(x, tv3, is_gx1_square)
+23.   y = CMOV(y, y1, is_gx1_square)
+24. tv1 = u mod 2
+25. tv2 = y mod 2
+26.   y = CMOV(-y, y, tv1 == tv2)s
+27.   x = x / tv4
+28. return (x, y)
+~~~
+
+This procedure depends on the following sqrt_ratio subroutine. It depends
+on the same value h defined above.
+
+~~~
+sqrt_ratio(u, v)
+
+Parameters:
+- F, a finite field of characteristic p and order q = p^m.
+
+Input: u and v, elements of F.
+Output: (b, y), where b is true and y = sqrt(u / v) if (u / v) is square in F,
+  and b is false and y = sqrt(h * (u / v)) otherwise.
```

Comment:
```suggestion
Output: (b, y), where b = True and y = sqrt(u / v) if (u / v) is square in F,
  and b = False and y = sqrt(S * (u / v)) otherwise.
```